# Pandas Complete tutorial for data science in 2025 



### Import

Before moving on to learn pandas first we need to install it and import it. If you install [Anaconda distributions](https://www.anaconda.com/) on your local machine or using [Google Colab](https://research.google.com/colaboratory) then pandas will already be available there, otherwise, you follow this installation process from [pandas official's website](https://pandas.pydata.org/docs/getting_started/install.html).

In [1]:
# Importing librariesdz
import numpy as np
import pandas as pd

# 2. Loading Different Data Formats Into a Pandas Data Frame

### Reading CSV file


In [14]:
# Import Dataset 
df=pd.read_csv("C:\\Users\\Techspot\\Documents\\New folder\\dataset\\online_store_customer_data.csv")

In [6]:
df

,Transaction_date,Transaction_ID,Gender,Age,Marital_status,State_names,Segment,Employees_status,Payment_method,Referal,Amount_spent
0,1/1/2019,151200,Female,19.0,Single,Kansas,Basic,Unemployment,Other,1.0,2051.36
1,1/1/2019,151201,Male,49.0,Single,Illinois,Basic,self-employed,Card,0.0,544.04
2,1/1/2019,151202,Male,63.0,Married,New Mexico,Basic,workers,PayPal,1.0,1572.60
3,1/1/2019,151203,NaN,18.0,Single,Virginia,Platinum,workers,Card,NaN,1199.79
4,1/1/2019,151204,Male,27.0,Single,Connecticut,Basic,self-employed,Card,0.0,NaN
...,...,...,...,...,...,...,...,...,...,...,...
2507,5/1/2021,153695,Female,57.0,Single,South Carolina,Platinum,self-employed,Card,0.0,150.10
2508,5/1/2021,153696,Female,36.0,Married,Hawaii,Silver,self-employed,PayPal,1.0,708.88
2509,5/1/2021,153697,Male,22.0,Single,South Carolina,Basic,workers,PayPal,1.0,2030.07
2510,5/1/2021,153698,NaN,44.0,Single,New York,Basic,Employees,PayPal,0.0,1909.77


In [16]:
df.head(6)

,Transaction_date,Transaction_ID,Gender,Age,Marital_status,State_names,Segment,Employees_status,Payment_method,Referal,Amount_spent
0,1/1/2019,151200,Female,19.0,Single,Kansas,Basic,Unemployment,Other,1.0,2051.36
1,1/1/2019,151201,Male,49.0,Single,Illinois,Basic,self-employed,Card,0.0,544.04
2,1/1/2019,151202,Male,63.0,Married,New Mexico,Basic,workers,PayPal,1.0,1572.60
3,1/1/2019,151203,NaN,18.0,Single,Virginia,Platinum,workers,Card,NaN,1199.79
4,1/1/2019,151204,Male,27.0,Single,Connecticut,Basic,self-employed,Card,0.0,NaN
5,1/3/2019,151205,Male,71.0,Single,Hawaii,Basic,Employees,PayPal,1.0,2922.66


In [18]:
df.tail(7)

,Transaction_date,Transaction_ID,Gender,Age,Marital_status,State_names,Segment,Employees_status,Payment_method,Referal,Amount_spent
2505,4/30/2021,153693,Male,54.0,Married,Alabama,Basic,Employees,Card,0.0,266.61
2506,4/30/2021,153694,Male,34.0,Single,Florida,Missing,Employees,Other,1.0,286.82
2507,5/1/2021,153695,Female,57.0,Single,South Carolina,Platinum,self-employed,Card,0.0,150.10
2508,5/1/2021,153696,Female,36.0,Married,Hawaii,Silver,self-employed,PayPal,1.0,708.88
2509,5/1/2021,153697,Male,22.0,Single,South Carolina,Basic,workers,PayPal,1.0,2030.07
2510,5/1/2021,153698,NaN,44.0,Single,New York,Basic,Employees,PayPal,0.0,1909.77
2511,5/1/2021,153699,Male,48.0,Single,California,Silver,workers,PayPal,1.0,1073.15


In [20]:
df.shape

(2512, 11)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2512 entries, 0 to 2511
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction_date  2512 non-null   object 
 1   Transaction_ID    2512 non-null   int64  
 2   Gender            2484 non-null   object 
 3   Age               2470 non-null   float64
 4   Marital_status    2512 non-null   object 
 5   State_names       2512 non-null   object 
 6   Segment           2512 non-null   object 
 7   Employees_status  2486 non-null   object 
 8   Payment_method    2512 non-null   object 
 9   Referal           2356 non-null   float64
 10  Amount_spent      2270 non-null   float64
dtypes: float64(3), int64(1), object(7)
memory usage: 216.0+ KB


In [ ]:
# Loading csv file with skip first 2 rows without header
df = pd.read_csv('C:\\Users\\Techspot\\Desktop\\PANDAS\\pandas-tutorial-article-2022-main\\pandas-tutorial-article main\\dataset\\online_store_customer_data.csv')

In [ ]:
# Loading csv file with skip first 2 rows without header
df_csv = pd.read_csv('C:\\Users\\Techspot\\Desktop\\PANDAS\\pandas-tutorial-article-2022-main\\pandas-tutorial-article main\\dataset\\online_store_customer_data.csv'
                     ,skiprows=3)
df_csv.head()

In [ ]:
pip install pandas

###### Read CSV file from URL

In [ ]:
# Read csv file from url
url1="https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
df_url = pd.read_csv(url1)
df_url

In [ ]:
#using UCI Repository
!pip install ucimlrepo

In [ ]:
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset
bank_marketing = fetch_ucirepo(id=222) 

# data (as pandas dataframes) 
X = bank_marketing.data.features 
y = bank_marketing.data.targets 

# metadata 
#print(bank_marketing.metadata) 
 
# variable information 
print(bank_marketing.variables) 


In [ ]:
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
iris = fetch_ucirepo(id=53) 
  
# data (as pandas dataframes) 
X = iris.data.features 
y = iris.data.targets 
  
# metadata 
print(iris.metadata) 
  
# variable information 
print(iris.variables) 


In [ ]:
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
heart_disease = fetch_ucirepo(id=45) 
  
# data (as pandas dataframes) 
X = heart_disease.data.features 
y = heart_disease.data.targets 
  
# metadata 
print(heart_disease.metadata) 
  
# variable information 
print(heart_disease.variables) 


### Write CSV file

In [ ]:
# saving df_url dataframe to csv file
df_url.to_csv('dataset/csv_from_url.csv')
df_url.to_csv('dataset/demo_text.txt')

### Read text file

In [ ]:
# read plain text file
df_txt = pd.read_csv("C:\\Users\\Techspot\\Documents\\New folder\\dataset\\demo_text.txt")
df_txt

### Read Excel file

In [ ]:
                    # read excel file
df_excel = pd.read_excel('C:\\Users\\Techspot\\Documents\\New folder\\dataset\\excel_file.xlsx', sheet_name='Sheet2')
df_excel

### Write Excel file

In [ ]:
# save dataframe to the excel file
df_url.to_csv('demo.xlsx')

## 3. Data preprocessing
Data preprocessing is the process of making raw data to clean data. This is the most crucial part of data the science. In this section, we will explore data first then we remove unwanted columns, remove duplicates, handle missing data, etc. After this step, we get clean data from raw data.

### 3.1 Data Exploring

#### 


Retrieving rows from data frame.

###### display first 6 rows
df.head(6)


In [ ]:
# display last 6 rows
df.tail(6)

#### Retrieving sample rows from data frame.



In [ ]:
# Display random 7 sample rows
df.sample(7)

#### Retrieving information about dataframe 

In [ ]:
# display datatypes
df.dtypes

In [ ]:
df.dtypes.value_counts()

#### Display number of rows and columns. 

In [ ]:
df.columns

In [ ]:
# display Age columndcdes first 6 rows data
df['Age'].head(6)

In [ ]:

# display first 4 rows of Age, Transaction_date and Gender columns
df[['Age', 'Transaction_date', 'Gender']].head(4)

In [ ]:
  # display last 10 rows of Age, Transaction_date and Gender columns
df[['Age', 'Transaction_date', 'Gender']].tail(10)

#### Retrieving a Range of Rows

In [ ]:
# for display 2nd to 6th rows
df[2:7]

# for display starting to 10th
df[:11]

# for display last two rows
df[-2:]


 ### 3.2 Data Cleaning
After the explore our datasets may need to clean them for better analysis. Data coming in from multiple sources so It's possible to have an error in some values. This is where data cleaning becomes extremely important. In this section, we will delete unwanted columns, rename columns, correct appropriate data types, etc.


#### Delete Columns name

In [ ]:
# Drop unwanted columns
df.drop(['Transaction_ID'], axis=1, inplace=True)

In [ ]:
df

#### Change Columns name

In [ ]:
# create new df_col dataframe from df.copy() method. 
df_col = df.copy()
df_col

In [ ]:
# rename columns name
df_col.rename(columns={"Transaction_date": "Date", "Gender": "Sex"}, inplace=True)
df_col.head(3)

#### Adding a new column to a DataFrame



In [ ]:
# Add a new ajusted column which value will be amount_spent + 10
df_col['Test'] = df_col['Amount_spent'] * 10

In [ ]:
df_col.head(3)

#### 
String value change or replace 

In [ ]:
df_col.head(3)

In [ ]:
# changing Female to Woman and Male to Man in Sex column.
#first argument in loc function is condition and second one is columns name. 
df_col.loc[df_col.Sex == "Female", 'Sex'] = 'Woman' 
df_col.loc[df_col.Sex == "Male", 'Sex'] = 'Man'

In [ ]:
df_col.head(3)

Now Sex columns values are changed Female to Woman and Male to Man.

#### Datatypes change

In [ ]:
df_col.info()

In our `Date` columns, it's object type so now we will convert this to date types, and also we will convert `Referal` columns float64 to float32.


In [ ]:

# change object type to datefime64 format
df_col['Date'] = df_col['Date'].astype('datetime64[ns]')

# change float64 to float32 of Referal columns
df_col['Referal'] = df_col['Referal'].astype('float32')


In [ ]:
df_col.info()

### 3.3 Remove duplicate

In [ ]:
# Display duplicated entries 
df_col.duplicated().sum()

In [ ]:
df_col.duplicated().head()

In [ ]:
# duplicate rows dispaly, keep arguments will--- 'first', 'last' and False
duplicate_value = df.duplicated(keep='first')
duplicate_value
df.loc[duplicate_value, :]

# dropping ALL duplicate values
df.drop_duplicates(keep = 'first', inplace = True)
df.drop_duplicates()|

### 3.4 Handling missing values

Handling missing values in the common task in the data pre-processing part. For many reasons most of the time we will encounter missing values. Without dealing with this we can't do the proper model building. For this section first, we will find out missing values then we decided how to handle them. We can handle this by removing affected columns or rows or replacing appropriate values there.

#### Display missing values information

In [ ]:
df.isna().sum().sort_values(ascending=False)

In [ ]:
df

#### Delete Nan rows



If we have less Nan value then we can delete entire rows by `dropna()` function. For this function, we will add columns name in subset parameter.

In [12]:
# df copy to df_copy
df_new = df.copy()

In [ ]:
df_new.info()

In [ ]:
#Delete Nan rows of Employees_status Columns
df_new.dropna(subset = ["Employees_status"], inplace=True)
df_new

#### Delete entire columns

If we have a large number of nan values in particular columns then dropping those columns might be a good decision rather than imputing.

In [ ]:
df_new.drop(columns=['Amount_spent'], inplace=True)
df_new

In [ ]:
df_new.isna().sum().sort_values(ascending=False)

#### Impute missing values

Sometimes if we delete entire columns that will be not the appropriate approach. Delete columns can affect our model building because we will lose our main features. For imputing we have many approaches so here are some of the most popular techniques.

**Method 1** - Impute fixed value like 0, 'Unknown' or 'Missing' etc. We inpute Unknown in Gender columns

In [ ]:
df

In [ ]:
df['Gender'].fillna('ABC', inplace=True)
df

**Method 2** - Impute Mean, Median and Mode

In [ ]:
# Impute Mean in Amount_spent columns
mean_amount_spent = df['Amount_spent'].mean()
df['Amount_spent'].fillna(mean_amount_spent, inplace=True)
df
mean_amount_spent

#Impute Median in Age column
median_age = df['Age'].median()
df['Age'].fillna(median_age, inplace=True)
median_age
df

# Impute Mode in Employees_status column
mode_emp = df['Employees_status'].mode().iloc[0]
df['Employees_status'].fillna(mode_emp, inplace=True)

In [ ]:
mean_amount_spent

In [ ]:
mode_emp

In [ ]:
median_age

In [ ]:
mode_emp

**Method 3** - Imputing forward fill or backfill by `ffill` and `bfill`. In `ffill` missing value impute from the value of the above row and for `bfill` it's taken from the below rows value.

In [ ]:
df

In [ ]:
df['Referal'].fillna(method='ffill', inplace=True)
df

In [ ]:
df.isna().sum().sum()

In [ ]:
df.isna().sum().sort_values(ascending=False)

Now we deal with all missing values with different methods. So now we haven't any null values.

## 4. Memory management

When we work on large datasets, There we get one big issue is a memory problem. We need too large resources for dealing with this. But there are some methods in pandas to deal with this. Here are some methods or strategies to deal with this problem with help of pandas.

### Change datatypes

In [ ]:
df_memory = df.copy()

In [ ]:
memory_usage = df_memory.memory_usage(deep=True)
memory_usage_in_mbs = round(np.sum(memory_usage / 1024 ** 2), 3)
print(f" Total memory taking df_memory dataframe is : {memory_usage_in_mbs:.2f} MB ")

#### Change object to category datatypes

Our data frame is small in size. Which is 1.15 MB. Now We will convert our object datatype to category.

In [ ]:
# Object datatype to category convert
df_memory[df_memory.select_dtypes(['object']).columns] = df_memory.select_dtypes(['object']).apply(lambda x: x.astype('category'))


In [ ]:
df_memory[df_memory.select_dtypes(['float64']).columns] = df_memory.select_dtypes(['float64']).apply(lambda x: x.astype('float32'))

In [ ]:
# convert object to category
df_memory.info(memory_usage="deep")

Now its reduce 1.15 megabytes to 216.6 kb. It's almost reduced 5.5 times.

#### Change int64 or float64 to int 32, 16, or 8

By default, pandas store numeric values to int64 or float64. Which takes more memory. If we have to store small numbers then we can change to 64 to 32, 16, and so on. For example, our Referal columns have only 0 and 1 values so for that we don't need to store at float64. so now we change it to float16.

In [ ]:
# Change Referal column datatypes
df_memory['Referal'] = df_memory['Referal'].astype('float16')

In [ ]:
# convert object to category
df_memory.info(memory_usage="deep")

After changing only one column's data types we reduce 216 kb to 179 kb.


**Note: Before changing datatypes please make sure it's consequences.**

## 5. Data Analysis

### 5.1. Calculating Basic statistical measurement

In [ ]:
df.describe().T

We know already above code will display only numeric columns basic statistical information. for object or category columns we can use `describe(include=object)`.

In [ ]:
df.describe(include=object).T

We can calculate the mean, median, mode, maximum values, minimum values of individual columns we simply use these functions.

In [ ]:
# Calculate Mean
mean = df['Age'].mean()

# Calculate Median
median = df['Age'].median()

#Calculate Mode
mode = df['Age'].mode().iloc[0]

# Calculate standard deviation
std = df['Age'].std()

# Calculate Minimum values
minimum = df['Age'].min()

# Calculate Maximum values
maximum = df.Age.max()

print(f" Mean of Age : {mean}")
print(f" Median of Age : {median}")
print(f" Mode of Age : {mode}")
print(f" Standard deviation of Age : {std:.2f}")
print(f" Maximum of Age : {maximum}")
print(f" Menimum of Age : {minimum}")

In pandas we can display the correlation of different numeric columns. For this we can use `.corr()` function.

### 5.2 Basic built in function for data analysis 

#### Number of uniqe values in category column

In [ ]:
# for display how many unique values are there in State_names column
df['State_names'].nunique()

#### Shows all unique values

In [ ]:
# for display uniqe values of State_names column
df['State_names'].unique()

#### Counts of unique values

In [ ]:
df['State_names'].value_counts().T

If we want to show with the percentage of occurrence rather number than we use `normalize=True` argument in `value_counts()` function

In [ ]:
# Calculate percentage of each category
df['State_names'].value_counts(normalize=True)

In [ ]:
df['State_names'].value_counts().sort_values(ascending = False).head(20)

In [ ]:
# Sort Values by State_names
df.sort_values(by=['State_names']).head(3)

For sorting our dataframe by Amount_spent with ascending order:

In [ ]:
# Sort Values Amount_spent with ascending order
df.sort_values(by=['Amount_spent']).head(5)

For sorting our dataframe by Amount_spent with descending order:

In [ ]:
# Sort Values Amount_spent with descending order
df.sort_values(by=['Amount_spent'], ascending=False).head(3)

Alternatively, We can use `nlargest()` and `nsmallest()` functions for displaying largest and smallest values with desired numbers. for example, If we want to display 4 largest Amount_spent rows then we use this:

In [ ]:
# nlargest
df.nlargest(2, 'Amount_spent').head() # first argument is how many rows you want to disply and second one is columns name

For 3 smallest Amount_spent rows

In [ ]:
# nsmallest
df.nsmallest(3, 'Age').head()